Update MPICH to 3.4.2

cticenhour opened the pull request:
This may help us avoid spurious MPI errors on ARM64 platforms (including Apple Silicon) that were occurring under 3.3.2. 

Updating to MPICH 3.4.2 requires the presence of `_SC_PHYS_PAGES`, which was first available in macOS 10.11 (El Capitan). Thus, updating the minimum deployment target from 10.9 (Mavericks) to 10.11 is required. 

Setting this to WIP until the proper SDKs are available on the build boxes. 

@dschwen could you give this a try on your M1 Mac Mini? I've only been building on Intel so far, but hope to get a build on M1 going tomorrow. 

Also tagging @milljm 

Refs #18954

cticenhour committed:
Update MPICH to 3.4.2

This is an attempt to avoid spurious MPI failures with Apple Silicon.

GLIBC >= 2.14 is required to build libmesh on Linux with this version of MPICH. Adding a new requirement for 2.17, as conda only has 2.12 and 2.17 available.

Refs #18954

cticenhour committed:
Require 10.11 as minimum macOS deployment target 

_SC_PHYS_PAGES support is required for MPICH 3.4.2, and was only 
available in 10.11 and later. 

refs #18954

cticenhour commented:
Currently seeing the following tests fail in `opt` on Intel: 

```
FAIL FAILED (CRASH) mesh/split_uniform_refine.square [min_cpus=3]
FAIL FAILED (CRASH) mesh/uniform_refine.square [min_cpus=3]
FAIL FAILED (CRASH) mesh/mesh_only.mesh_only_checkpoint [min_cpus=3]
FAIL FAILED (CRASH) meshgenerators/distributed_rectilinear/generator.3d_scomm_out [min_cpus=3]
```

All have the same `vtkMPICommunicator` related error message. A representative example is below:

```
mesh/mesh_only.mesh_only_checkpoint: 2021-11-22 19:50:14.147 (   0.207s) [          3E614F] vtkMPICommunicator.cxx:67    WARN| MPI had an error
mesh/mesh_only.mesh_only_checkpoint: ------------------------------------------------
mesh/mesh_only.mesh_only_checkpoint: Invalid datatype, error stack:
mesh/mesh_only.mesh_only_checkpoint: MPI_Test(188)............: MPI_Test(request=0x7fe980004788, flag=0x7ffee6ca7d0c, status=0x1) failed
mesh/mesh_only.mesh_only_checkpoint: MPIR_Test(81)............:
mesh/mesh_only.mesh_only_checkpoint: MPIDIG_recv_copy_seg(220): Receiving data with a datatype whose signature does not match that of the sending datatype.
```

milljm commented:
The machines are ready to use 10.11 SDK.

dschwen commented:
> @dschwen could you give this a try on your M1 Mac Mini?

I'm on PL until next week though

cticenhour commented:
@dschwen thanks for the reply while away. It'll keep until next week - I have a few bugs to sort out still on the Intel Mac and Linux side.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/899346/) on 04f3f64 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19456/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Create Mac Stack](https://civet.inl.gov/job/899319/) on 04f3f64 : invalidated by @milljm

new sandbox change, try again (allowing access to required SDK path)

moosebuild commented:
Job [Create Mac Stack](https://civet.inl.gov/job/899756/) on f723d77 : invalidated by @milljm

use conda build instead of mambabuild when determining what package was produced, and where

dschwen commented:
Hmmm, I did `conda build mpich`, which worked fine, but then `conda build petsc` gave me

```
                ERROR in COMMAND LINE ARGUMENT to ./configure
-------------------------------------------------------------------------------
The option -I/Users/schwd/miniforge3/envs/build/conda-bld/moose-petsc_1638152792665/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/include should probably be -I/Users/schwd/miniforge3/envs/build/conda-bld/moose-petsc-1638152792665/-h-env-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placehold-placeho/include
```

milljm commented:
@dschwen I don't think the fix for this has entered this branch yet. Indeed, on the M1, I have needed to quote the incoming arguments:

```diff
diff --git a/scripts/configure_petsc.sh b/scripts/configure_petsc.sh
index ee7d13b197..dea877fc34 100644
--- a/scripts/configure_petsc.sh
+++ b/scripts/configure_petsc.sh
@@ -44,7 +44,7 @@ function configure_petsc()
       --with-fortran-bindings=0 \
       --with-sowing=0 \
       --with-64-bit-indices \
-      $*
+      "$*"
 
   return $?
 }
```

Not sure if this is your issue...

cticenhour commented:
I definitely should add that fix to this branch. Will do that when I can get back to the Mac side of things. Linux dependency issues are still holding this back for the moment.

cticenhour committed:
Fixup VTK dependency libraries to be compatible with updated sysroot

The previous CDT packages default to CentOS 6, and merely using the
proper sysroot version in the package build requirements (as suggested
via conda-forge docs) is not enough to prompt usage of CentOS 7
compatible packages.

refs #18954

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/903176/) on 34e40d7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19456/site)

This comment will be updated on new commits.

milljm commented on conda/libmesh-vtk/meta.yaml:
```diff
@@ -20,32 +20,30 @@ build:
+    - libice-devel-cos7-x86_64          # [linux]
+    - libxext-devel-cos7-x86_64         # [linux]
+    - mesa-libgl-devel-cos7-x86_64      # [linux]
+    - xorg-x11-proto-devel-cos7-x86_64  # [linux]
```
You don't suppose these being updated with `mamba update --all` will break an installation of moose-libmesh-vtk? We may need to pin these like we do with {{ moose_mpich }}. Not sure though. Maybe allowing these libraries to update (without a new build of moose-libmesh-vtk) is fine...

cticenhour replied:
Pinning `moose_sysroot` via `moose-mpich` as we do should pin them (one of the versions here is 2018, so I don't think they'll be updated again). But would you like me to make the change? I think this fix will be temporary as conda-forge moves to a CentOS 7 default late this year into next.

cticenhour replied:
Right, that's how it normally works. These CDT packages are meant to be very stable, but I also see sense in making things a bit more foolproof. I'll add it in the next round of commits. :D

milljm replied:
lol, well this may all be moot, cause perhaps we don't need to be so bound to a specific version for VTK runtime?

cticenhour replied:
Since this CDT bit of conda is entirely new to me, I don't have a solid intuition on it. Perhaps better to be a bit conservative here since we know the current versions work. I would like to return to the more general syntax when we move to miniforge and conda-forge updates their minimums.

cticenhour commented:
@roystgnr I'm seeing an MPI error related to TIMPI that I wanted to put on your radar as I explore this a bit more. Took me a bit to find it, since the initial CIVET errors for this PR seemed to be related to VTK (and there may still be come issues there). Building libmesh manually without VTK support still had some parallel failures, seemingly related to distributed mesh communication and checkpoint restart. A representative example is the `mesh_only_checkpoint` test, where the failure output is:

```
Assertion `timpi_mpi_error_code == MPI_SUCCESS' failed.
timpi_mpi_error_code = 335675395
MPI_SUCCESS = 0
Invalid datatype, error stack:
MPI_Test(188)............: MPI_Test(request=0x6000002a8ee8, flag=0x7ff7b147d264, status=0x1) failed
MPIR_Test(81)............:
MPIDIG_recv_copy_seg(220): Receiving data with a datatype whose signature does not match that of the sending datatype.

[0] ../../../../contrib/timpi/src/parallel/src/request.C, line 157, compiled Dec 22 2021 at 11:16:33
Assertion `timpi_mpi_error_code == MPI_SUCCESS' failed.
timpi_mpi_error_code = 134348803
MPI_SUCCESS = 0
Invalid datatype, error stack:
MPI_Test(188)............: MPI_Test(request=0x6000005c5f08, flag=0x7ff7b257a264, status=0x1) failed
MPIR_Test(81)............:
MPIDIG_recv_copy_seg(220): Receiving data with a datatype whose signature does not match that of the sending datatype.

[1] ../../../../contrib/timpi/src/parallel/src/request.C, line 157, compiled Dec 22 2021 at 11:16:33


*** ERROR ***


Abort(1) on node 1 (rank 1 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1


*** ERROR ***


Abort(1) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

if run with `-j2` with the new set of packages on this PR. Reproducing this outside of TestHarness can be done with:

```
mpiexec -n 2 ../../../moose_test-devel -i mesh_only.i Mesh/parallel_type=distributed --mesh-only
```

and a backtrace on Rank 0 produced:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00000001198f70d8 libpmpi.12.dylib`MPIDI_POSIX_mpi_release_gather_gather + 616
    frame #1: 0x00000001198f7dbe libpmpi.12.dylib`MPIDI_POSIX_mpi_allreduce_release_gather + 1182
    frame #2: 0x00000001198f4fe8 libpmpi.12.dylib`MPIR_Allreduce + 1864
    frame #3: 0x0000000110efb01c libmpi.12.dylib`MPI_Allreduce + 3084
    frame #4: 0x000000010f8eade7 libtimpi_devel.6.dylib`void TIMPI::Communicator::min<unsigned long>(this=<unavailable>, r=<unavailable>) const at parallel_implementation.h:2057:7 [opt]
    frame #5: 0x000000010f8e9b92 libtimpi_devel.6.dylib`void TIMPI::Communicator::min<short, std::__1::allocator<short> >(std::__1::vector<short, std::__1::allocator<short> >&) const at parallel_implementation.h:1941:13 [opt]
    frame #6: 0x000000010f8e9b78 libtimpi_devel.6.dylib`void TIMPI::Communicator::min<short, std::__1::allocator<short> >(this=0x0000600000936a48, r=size=32) const at parallel_implementation.h:2073:7 [opt]
    frame #7: 0x000000010f8e63b2 libtimpi_devel.6.dylib`bool TIMPI::Communicator::verify<std::__1::vector<short, std::__1::allocator<short> > >(this=0x0000600000936a48, r=size=32) const at parallel_implementation.h:1941:13 [opt]
    frame #8: 0x000000010f8e6228 libtimpi_devel.6.dylib`TIMPI::Communicator::verify(this=<unavailable>, r=<unavailable>) const at communicator.C:330:20 [opt]
    frame #9: 0x00000001161de00e libmesh_devel.0.dylib`libMesh::MeshCommunication::broadcast(this=<unavailable>, mesh=0x00007fdda69068b0) const at mesh_communication.C:1164:3 [opt]
    frame #10: 0x0000000113d653a2 libmoose-devel.0.dylib`FileMeshGenerator::generate(this=0x0000600001d20018) at FileMeshGenerator.C:78:27 [opt]
    frame #11: 0x0000000113dd0b22 libmoose-devel.0.dylib`MeshGenerator::generateInternal(this=0x0000600001d20018) at MeshGenerator.C:120:15 [opt]
    frame #12: 0x00000001148ee69e libmoose-devel.0.dylib`MooseApp::executeMeshGenerators(this=<unavailable>) at MooseApp.C:2055:38 [opt]
    frame #13: 0x000000011437fc13 libmoose-devel.0.dylib`Action::timedAct(this=0x00007fdda9705178) at Action.C:93:3 [opt]
    frame #14: 0x000000011437f5c9 libmoose-devel.0.dylib`ActionWarehouse::executeActionsWithAction(this=0x00007fdda885c148, task="execute_mesh_generators") at ActionWarehouse.C:384:19 [opt]
    frame #15: 0x00000001143bad08 libmoose-devel.0.dylib`ActionWarehouse::executeAllActions(this=0x00007fdda885c148) at ActionWarehouse.C:344:5 [opt]
    frame #16: 0x00000001148d46f5 libmoose-devel.0.dylib`MooseApp::runInputFile(this=0x00007fdda885ba18) at MooseApp.C:1021:21 [opt]
    frame #17: 0x00000001148ce020 libmoose-devel.0.dylib`MooseApp::run(this=0x00007fdda885ba18) at MooseApp.C:1450:5 [opt]
    frame #18: 0x000000010f76dc34 moose_test-devel`main(argc=6, argv=0x00007ff7b0797d80) at main.C:33:8 [opt]
    frame #19: 0x000000011392c4fe dyld`start + 462
```

I'll be taking a look at `parallel_implementation` shortly, but as I have relatively little MPI communication experience, I welcome any insight you might have.

cticenhour commented:
@roystgnr pinging you again in case anything I posted above right before the holidays could be useful to you. Probably something you and @loganharbour have probably seen already, just wanted to make sure.

loganharbour commented:
Yeah, luckily it's the same issue.

roystgnr commented:
Some basic questions:

Has anyone tried 3.4.3 yet?  There's a change note about fixing a leak in predefined datatypes, which doesn't sound like it could be the cause, but ...

Are you guys building mpich with OFI or UCX?  My workstation doesn't have anything fancy and MPICH apparently refuses to select a basic network device interface by default these days.

cticenhour commented:
At the time this PR was made, 3.4.3 did not yet exist, and I haven't had the chance to try it since being back from the holiday. I should give it a go. 

For ARM MacOS and Linux, we have selected the legacy option (CH3) instead of the new default (CH4, with OFI or UCX able to be built by the installer), but for Intel MacOS I was experimenting with the new default here. I'll attempt a build with 3.4.3 with the current settings, and another with legacy communication settings turned on for everything.

roystgnr commented:
Wait a minute ... is this *only* failing in the Mac tests?  I was under the impression that Logan had managed to trigger it on Linux too.

cticenhour commented:
Keep in mind that I modified the Linux dependencies as well as the communication setting to pass the Linux tests. I cannot remember the results using the new default, so there might still be breakages to fix if that setting is unified for all platforms.

cticenhour commented:
Seeing the same failures with 3.4.3 / CH4 on Mac.....in the process of building 3.4.3 / CH3. I can re-try the Linux build with CH4 on rod to see where things stand there.

loganharbour commented:
I triggered it on a spack built MPICH 3.4.2.

cticenhour commented:
I am not triggering this issue using CH3 with 3.4.3 on Mac. Just for sanity's sake I should double-check this on 3.4.2 as well.

cticenhour commented:
After discussing further with @loganharbour, I am setting the MPICH abstract device interface (ADI) to the legacy option (CH3) for all platforms for now (which was the default in MPICH 3.3.x anyway). Future work will need to be done to allow an update to the new device default (CH4).

cticenhour committed:
Update moose newsletter (#19456)

milljm reviewed (approved):
🥁
